streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -673,9 +673,15 @@ private void checkBrokerVersionCompatibility() throws StreamsException {
     public synchronized void start() throws IllegalStateException, StreamsException {
         log.debug("{} Starting Streams client", logPrefix);
 
+        final State oldState = state();
+        final State runningState = State.RUNNING;
+        // this is a mandatory check and not covered by regular state transition checks (cf. KAFKA-5818)
+        if (!(oldState == State.CREATED)) {
```

Comment:
Hmm.. what if after the check passed, i.e. between line 681 and 682, the state is then transited to running and then to rebalancing?

Reply:
I think we can introduce a

```
boolean setStateFromCreated()
```

in which we grab the state lock, and then check that state is `CREATED`, and then call `setState(RUNNING)` assuming that `sychronized` is reentrant.

Reply:
+1

Reply:
Ack.
We can't call `setState` though as we would call the stateListener while we hold the lock...